examples/jsm/controls/OrbitControls.js

```diff
@@ -271,6 +269,77 @@ class OrbitControls extends EventDispatcher {
 
 				}
 
+				// adjust camera position
+				let zoomChanged = false;
+				if ( scope.zoomToCursor ) {
+
+					let newRadius;
+					if ( scope.object.isPerspectiveCamera ) {
+
+						// move the camera down the pointer ray
+						// this method avoids floating point error
+						const prevRadius = offset.length();
+						newRadius = prevRadius * scale;
+						newRadius = Math.max( scope.minDistance, Math.min( scope.maxDistance, newRadius ) );
+
+						const radiusDelta = prevRadius - newRadius;
+						scope.object.position.addScaledVector( dollyDirection, radiusDelta );
+
+					} else if ( scope.object.isOrthographicCamera ) {
+
+						// adjust the ortho camera position based on zoom changes
+						const mouseBefore = new Vector3( mouse.x, mouse.y, 0 );
+						mouseBefore.unproject( scope.object );
+
+						scope.object.zoom = Math.max( scope.minZoom, Math.min( scope.maxZoom, scope.object.zoom / scale ) );
+						scope.object.updateProjectionMatrix();
+						zoomChanged = true;
+
+						const mouseAfter = new Vector3( mouse.x, mouse.y, 0 );
+						mouseAfter.unproject( scope.object );
+
+						scope.object.position.sub( mouseAfter ).add( mouseBefore );
+						newRadius = offset.length();
+
+					} else {
+
+						newRadius = offset.length();
+
+					}
+
+					scope.object.updateMatrixWorld();
+
+					if ( this.screenSpacePanning ) {
+
+						// position the orbit target in front of the new camera position
+						scope.target.set( 0, 0, - 1 )
+							.transformDirection( scope.object.matrix )
+							.multiplyScalar( newRadius )
+							.add( scope.object.position );
+
+					} else {
+
+						// get the ray and translation plane to compute target
+						const ray = new Ray();
```

Comment:
How about caching this `ray`? I.e. creating it only once

Reply:
Yeah I plan to cache both ray and plane once understanding how to handle the corner case mentioned above.